plugins/inputs/ping_native/ping.go

```diff
@@ -0,0 +1,260 @@
+
+	network    string                 // network is the network to listen on ("udp4", "udp6", "ip4:icmp", "ip6:ip6-icmp")
+	listenAddr string                 // listenAddr is the address associated with the interface defined.
+	hostCache  []net.Addr             // hosts to ping
+	hosts      map[string]string      // ip to host map
+	rcvdCache  map[string]pingResults // cache of echo responses received
+}
+
+func (*Ping) Description() string {
+	return "Ping given host(s) and return statistics"
+}
+
+const sampleConfig = `
+  ## List of hosts to ping.
+  hosts = ["8.8.8.8"]
+
+  ## Number of pings to send per collection.
+  # count = 1
+
+  ## Interval, in s, at which to ping.
+  # ping_interval = 1.0
+
+  ## Per-ping timeout, in s (0 == no timeout).
+  # timeout = 1.0
+
+  ## Total-ping deadline, in s. Set to value equal to or lower than agent interval.
+  # deadline = 10
+
+  ## Interface or source address to send ping from.
+  # interface = ""
+
+	## Whether to ping ipv6 addresses.
+  # ipv6 = false
+`
+
+func (*Ping) SampleConfig() string {
+	return sampleConfig
+}
+
+func getAddr(iface string) string {
+	if addr := net.ParseIP(iface); addr != nil {
+		return addr.String()
+	}
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return ""
+	}
+
+	var ip net.IP
+	for i := range ifaces {
+		if ifaces[i].Name == iface {
+			addrs, err := ifaces[i].Addrs()
+			if err != nil {
+				return ""
+			}
+			if len(addrs) > 0 {
+				switch v := addrs[0].(type) {
+				case *net.IPNet:
+					ip = v.IP
+				case *net.IPAddr:
+					ip = v.IP
+				}
+				if len(ip) == 0 {
+					return ""
+				}
+				return ip.String()
+			}
+		}
+	}
+
+	return ""
+}
+
+func (p *Ping) buildHostCache(acc telegraf.Accumulator) {
+	if p.hostCache != nil {
+		return
+	}
+
+	p.network = "ip4:icmp"
+	if runtime.GOOS == "darwin" || runtime.GOOS == "linux" {
+		p.network = "udp4"
+		if p.IPV6 {
+			p.network = "udp6"
+		}
+	} else if p.IPV6 {
+		p.network = "ip6:ipv6-icmp"
+	}
+
+	p.hostCache = []net.Addr{}
+	p.hosts = map[string]string{}
+
+	for _, host := range p.Hosts {
+		var addr net.Addr
+		var err error
+		if p.IPV6 {
+			addr, err = net.ResolveIPAddr("ip6", host)
+		} else {
+			addr, err = net.ResolveIPAddr("ip4", host)
+		}
+
+		switch runtime.GOOS {
+		case "js", "nacl", "plan9", "windows":
+			p.hosts[addr.String()] = host
+		default:
+			if p.IPV6 {
+				p.hosts["["+addr.String()+"]:0"] = host
+			} else {
+				p.hosts[addr.String()+":0"] = host
+			}
+		}
+
+		if a, ok := addr.(*net.IPAddr); ok && runtime.GOOS == "darwin" || runtime.GOOS == "linux" {
+			addr = &net.UDPAddr{IP: a.IP, Zone: a.Zone}
+		}
+
+		if err != nil {
+			acc.AddFields("ping", map[string]interface{}{"result_code": 1}, map[string]string{"source": host})
+			acc.AddError(err)
+			continue
+		}
+		if addr == nil {
+			continue
+		}
+		p.hostCache = append(p.hostCache, addr)
+	}
+}
+
+func (p *Ping) onRecv(pkt *ping.Packet) {
+	p.rcvdCache[pkt.IPAddr] = pingResults{ttl: pkt.TTL}
```

Comment:
I think you would need to lock this, maybe it can be done without a member variable.